domain/src/main/java/org/sagebionetworks/research/domain/task/navigation/TreeNavigator.java

```diff
@@ -0,0 +1,336 @@
+     * contains the progress marker at this index.
+     */
+    private int getLastIndexInMarkers(List<String> identifiers) {
+        if (this.progressMarkers != null) {
+            for (int i = this.progressMarkers.size() - 1; i >= 0; i--) {
+                if (identifiers.contains(progressMarkers.get(i))) {
+                    return i;
+                }
+            }
+        }
+
+        return -1;
+    }
+
+    /*
+     * A class to represent a single node in the TreeNavigator
+     */
+    private static final class Node {
+        // The step that this node represents
+        @Nullable
+        private final Step step;
+
+        // The children if any that this node has. children.isEmpty() == false always, when
+        // children == null the node is a leaf.
+        @Nullable
+        private final ImmutableList<Node> children;
+
+        /**
+         * Constructs a new node from the given list of steps. This node will represent the root
+         * of a Tree in which the steps are the children.
+         * @param steps The list of steps to construct the node from.
+         */
+        private Node(@Nullable List<Step> steps) {
+            this.step = null;
+            this.children = constructChildNodes(steps);
+        }
+
+        /**
+         * Constructs a new node corresponding to the given step.
+         * @param step the step to construct the node from.
+         */
+        private Node(@NonNull Step step) {
+            this.step = step;
+            if (step instanceof SectionStep) {
+                this.children = constructChildNodes(((SectionStep)step).getSteps());
+            } else {
+                this.children = null;
+            }
+        }
+
+        /**
+         * Constructs nodes from all the steps in the given list and returns them
+         * in an ImmutableList.
+         * @param childSteps the list of steps to construct nodes from.
+         * @return An ImmutableList of nodes constructed from the given steps.
+         */
+        private ImmutableList<Node> constructChildNodes(@Nullable List<Step> childSteps) {
+            if (childSteps != null && !childSteps.isEmpty()) {
+                List<Node> children = new ArrayList<>();
+                for (Step childStep : childSteps) {
+                    children.add(new Node(childStep));
+                }
+
+                ImmutableList.Builder<Node> builder = new ImmutableList.Builder<>();
+                builder.addAll(children);
+                return builder.build();
+            }
+
+            return null;
+        }
+
+
+        /**
+         * Returns true if this node is a leaf, false otherwise.
+         * @return true if this node is a leaf, false otherwise.
+         */
+        private boolean isLeaf() {
+            return this.children == null;
+        }
+
+        @Override
+        public String toString() {
+            if (this.step != null) {
```

Comment:
Should nodes print their children?

Reply:
I was using this so the Android Studio debugger would show the nodes in a more readable way. I left the children out to avoid having the text be too long.